component/newstorage/batchedstore/batchedstore.go

```diff
@@ -0,0 +1,380 @@
+// If the store cannot be found, then an error wrapping ErrStoreNotFound will be returned by the underlying provider.
+// If name is blank, then an error will be returned by the underlying provider.
+func (p *Provider) GetStoreConfig(name string) (newstorage.StoreConfiguration, error) {
+	config, err := p.underlyingProvider.GetStoreConfig(name)
+	if err != nil {
+		return newstorage.StoreConfiguration{},
+			fmt.Errorf("failed to get store config from underlying provider: %w", err)
+	}
+
+	return config, nil
+}
+
+// GetOpenStores returns all currently open stores.
+func (p *Provider) GetOpenStores() []newstorage.Store {
+	openStores := make([]newstorage.Store, len(p.openStores))
+
+	var counter int
+
+	for _, openStore := range p.openStores {
+		openStores[counter] = openStore
+		counter++
+	}
+
+	return openStores
+}
+
+// Close closes all stores created under this store provider.
+// For persistent store implementations, this does not delete any data in the underlying databases.
+func (p *Provider) Close() error {
+	var openStoresSnapshot []*store
+
+	p.Lock()
+	for _, openStore := range p.openStores {
+		openStoresSnapshot = append(openStoresSnapshot, openStore)
+	}
+	p.Unlock()
+
+	for _, openStore := range openStoresSnapshot {
+		err := openStore.Close()
+		if err != nil {
+			return fmt.Errorf(`failed to close open store with name "%s": %w`, openStore.name, err)
+		}
+	}
+
+	err := p.underlyingProvider.Close()
+	if err != nil {
+		return fmt.Errorf("failed to close underlying provider: %w", err)
+	}
+
+	return nil
+}
+
+// Flush forces all open stores to flush their data immediately.
+func (p *Provider) Flush() error {
+	p.Lock()
+	defer p.Unlock()
+
+	for _, openStore := range p.openStores {
+		openStore.Lock()
+
+		err := openStore.flush()
+		if err != nil {
+			return fmt.Errorf(`failed to flush data in open store with name "%s": %w`, openStore.name, err)
+		}
+
+		openStore.Unlock()
+	}
+
+	return nil
+}
+
+func (p *Provider) removeStore(name string) {
+	p.Lock()
+	defer p.Unlock()
+
+	delete(p.openStores, name)
+}
+
+type store struct {
+	name            string
+	underlyingStore newstorage.Store
+	currentBatch    []newstorage.Operation
+	batchSizeLimit  int
+	close           closer
+	*sync.RWMutex
+}
+
+func (s *store) Put(key string, value []byte, tags ...newstorage.Tag) error {
+	if key == "" {
+		return errEmptyKey
+	}
+
+	if value == nil {
+		return errors.New("value cannot be nil")
+	}
+
+	s.Lock()
+	defer s.Unlock()
+
+	s.currentBatch = append(s.currentBatch, newstorage.Operation{
+		Key:   key,
+		Value: value,
+		Tags:  tags,
+	})
+
+	if len(s.currentBatch) >= s.batchSizeLimit {
+		err := s.flush()
+		if err != nil {
+			return fmt.Errorf(failFlush, err)
+		}
+	}
+
+	return nil
+}
+
+// TODO (#2485): Check data in current batch before resorting to a flush.
+func (s *store) Get(key string) ([]byte, error) {
+	s.Lock()
+	defer s.Unlock()
+
+	if len(s.currentBatch) > 0 {
```

Comment:
I would put this check inside of Flush since it's repeated in multiple function. Also, you're acquiring a write lock each time you check currentBatch. The check should be done using a read lock and then use a write lock when modifying the value.

Reply:
Done.